config/module.go

```diff
@@ -115,46 +115,97 @@ func (m Module) NeedsSyntheticPackage() bool {
+		name = m.Name
+	}
+	return PackageConfig{Name: name, Package: name, Type: PackageTypeModule}, nil
 }
 
-// syntheticPackageExeDir returns the unpacked ExePath for local tarball modules.
-func (m Module) syntheticPackageExeDir(packagesDir string) (string, error) {
+// exeDir returns the parent directory for the unpacked module.
+func (m Module) exeDir(packagesDir string) (string, error) {
+	if !m.NeedsSyntheticPackage() {
+		return filepath.Dir(m.ExePath), nil
+	}
 	pkg, err := m.SyntheticPackage()
 	if err != nil {
 		return "", err
 	}
 	return pkg.LocalDataDirectory(packagesDir), nil
 }
 
-// EvaluateExePath returns absolute ExePath except for local tarballs where it looks for side-by-side meta.json.
-// The side-by-side lookup is because we don't bundle entrypoint in module tarballs, it's not an intentional design choice.
+// parseJSONPath returns a *T by parsing the json file at `path`.
+func parseJSONPath[T any](path string) (*T, error) {
+	f, err := os.Open(path) //nolint:gosec
+	if err != nil {
+		return nil, errors.Wrap(err, "reading json file")
+	}
+	var target T
+	err = json.NewDecoder(f).Decode(&target)
+	if err != nil {
+		return nil, err
+	}
+	return &target, nil
+}
+
+// EvaluateExePath returns absolute ExePath from one of three sources (in order of precedence):
+// 1. if there is a meta.json in the exe dir, use that, except in local non-tarball case.
+// 2. if this is a local tarball and there's a meta.json next to the tarball, use that.
+// 3. otherwise use the exe path from config, or fail if this is a local tarball.
+// Note: the working directory must be the unpacked tarball directory or local exec directory.
 func (m Module) EvaluateExePath(packagesDir string) (string, error) {
+	if !filepath.IsAbs(m.ExePath) {
+		return "", fmt.Errorf("expected ExePath to be absolute path, got %s", m.ExePath)
+	}
+	exeDir, err := m.exeDir(packagesDir)
+	if err != nil {
+		return "", err
+	}
+	// note: we don't look at internal meta.json in local non-tarball case because user has explicitly requested a binary.
+	localNonTarball := m.Type == ModuleTypeLocal && !m.NeedsSyntheticPackage()
+	if !localNonTarball {
+		// this is case 1, meta.json in exe folder.
+		metaPath, err := utils.SafeJoinDir(exeDir, "meta.json")
+		if err != nil {
+			return "", err
+		}
+		_, err = os.Stat(metaPath)
+		if err == nil {
+			// this is case 1, meta.json in exe dir
+			meta, err := parseJSONPath[JSONManifest](metaPath)
+			if err != nil {
+				return "", err
+			}
+			entrypoint, err := utils.SafeJoinDir(exeDir, meta.Entrypoint)
+			if err != nil {
+				return "", err
+			}
+			return filepath.Abs(entrypoint)
+		}
+	}
 	if m.NeedsSyntheticPackage() {
-		metaPath := filepath.Join(filepath.Dir(m.ExePath), "meta.json")
-		f, err := os.Open(metaPath) //nolint:gosec
+		// this is case 2, side-by-side
+		// TODO(RSDK-7848): remove this case once java sdk supports internal meta.json.
+		metaPath, err := utils.SafeJoinDir(filepath.Dir(m.ExePath), "meta.json")
```

Comment:
Have we documented how to use side-by-side? Wondering if someone external could be using it, therefore need to support it for longer past java SDK changes.

Reply:
it's used inside the mod.sh logic of the java sdk, so if someone is using java sdk 0.0.2 on android, they implicitly need side-by-side

otherwise we don't advertise this externally

Reply:
oops this is wrong, side-by-side meta is still used in java sdk 0.0.3 -- this is a remaining todo